src/core/utils/lookup_table.hpp

```diff
@@ -0,0 +1,149 @@
+     * @tparam kLength     The array length (number of entries in the array).
+     *
+     * @note In the common use of this method as `IsSorted(sTable)` where sTable` is a fixed size array, the template
+     * types/parameters do not need to be explicitly specified and can be deduced from the passed-in argument.
+     *
+     * @param[in] aTable  A reference to an array of `kLength` entries on type `EntryType`
+     *
+     * @retval TRUE   If the entries in @p aTable are sorted (alphabetical order).
+     * @retval FALSE  If the entries in @p aTable are not sorted.
+     *
+     */
+    template <class EntryType, uint16_t kLength> static constexpr bool IsSorted(const EntryType (&aTable)[kLength])
+    {
+        return IsSorted(&aTable[0], kLength);
+    }
+
+    /**
+     * This template method performs binary search in a given sorted table array to find an entry matching a given name.
+     *
+     * @note This method requires the array to be sorted, otherwise its behavior is undefined.
+     *
+     * @tparam EntryType   The table entry type. The `EntryType` MUST provide `mName` member variable as a `const
+     *                     char *`. For example, this can be realized by `EntryType` being a subclass of `Entry`.
+     * @tparam kLength     The array length (number of entries in the array).
+     *
+     * @note In the common use of this method as `Find(name, sTable)` where sTable` is a fixed size array, the template
+     * types/parameters do not need to be explicitly specified and can be deduced from the passed-in argument.
+     *
+     * @param[in] aName   A name string to search for within the table.
+     * @param[in] aTable  A reference to an array of `kLength` entries on type `EntryType`
+     *
+     * @returns A pointer to the entry in the table if a match is found, otherwise nullptr (no match in table).
+     *
+     */
+    template <class EntryType, uint16_t kLength>
+    static const EntryType *Find(const char *aName, const EntryType (&aTable)[kLength])
+    {
+        return reinterpret_cast<const EntryType *>(
+            Find(aName, &aTable[0], kLength, sizeof(aTable[0]), GetName<EntryType>));
+    }
+
+private:
+    typedef const char *(&NameGetter)(const void *aPointer);
+
+    template <class EntryType> static const char *GetName(const void *aPointer)
+    {
+        return reinterpret_cast<const EntryType *>(aPointer)->mName;
+    }
+
+    template <class EntryType> static constexpr bool IsSorted(const EntryType *aTable, uint16_t aLength)
+    {
+        return (aLength <= 1) ? true
+                              : AreInOrder(aTable[0].mName, aTable[1].mName) && IsSorted(aTable + 1, aLength - 1);
+    }
+
+    constexpr static bool AreInOrder(const char *aFirst, const char *aSecond)
+    {
+        return (*aFirst < *aSecond) ? true : ((*aFirst > *aSecond) ? false : AreInOrder(aFirst + 1, aSecond + 1));
```

Comment:
Do we properly handle strings that are equal? In other words, what if two adjacent strings are the same?

Reply:
Good point. I will update it (will add a case for it in unit test as well).

Reply:
Added this on new push (we ensure that there are no duplicate names in the table).